File: virtool/workflow/analysis/skewer.py
"""Utilities and a fixture for `Skewer <https://github.com/relipmoc/skewer>`_."""

import asyncio
import os
import shutil
from asyncio.subprocess import Process
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from tempfile import mkdtemp
from typing import Protocol

from pyfixtures import fixture

from virtool.models.enums import LibraryType
from virtool.workflow import RunSubprocess
from virtool.workflow.analysis.utils import ReadPaths
from virtool.workflow.data.samples import WFSample


class SkewerMode(str, Enum):
    """The mode to run Skewer in."""

    PAIRED_END = "pe"
    """Run Skewer in paired-end mode."""

    SINGLE_END = "any"
    """Run Skewer in single-end mode."""


@dataclass
class SkewerConfiguration:
    """A configuration for running Skewer."""

    min_length: int
    """The minimum length of a trimmed read."""

    mode: SkewerMode
    """The mode to run Skewer in."""

    end_quality: int = 20
    """The minimum quality score for the end of a trimmed read."""

    max_error_rate: float = 0.1
    """
    The maximum error rate for a trimmed read. Reads that exceed the rate will be
    discarded.
    """

    max_indel_rate: float = 0.03
    """
    The maximum indel rate for a trimmed read. Reads that exceed the rate will be
    discarded.
    """

    mean_quality: int = 25
    """The minimum mean quality score for a trimmed read. Reads  """

    number_of_processes: int = 1
    """The number of processes to use when running Skewer."""

    quiet: bool = True
    """Whether to run Skewer in quiet mode."""

    other_options: tuple[str] = ("-n", "-z")
    """Other options to pass to Skewer."""


@dataclass
class SkewerResult:
    """Represents the result of running Skewer to trim a paired or unpaired FASTQ dataset."""

    command: list[str]
    """The command used to run Skewer."""

    output_path: Path
    """The path to the directory containing the trimmed reads."""

    process: Process
    """The process that ran Skewer."""

    read_paths: ReadPaths
    """The paths to the trimmed reads."""

    @property
    def left(self) -> Path:
        """The path to one of:
        - the FASTQ trimming result for an unpaired Illumina dataset
        - the FASTA trimming result for the left reads of a paired Illumina dataset

        """
        return self.read_paths[0]

    @property
    def right(self) -> Path | None:
        """The path to the rights reads of a paired Illumina dataset.

        ``None`` if the dataset in unpaired.

        :type: :class:`.Path`

        """
        try:
            return self.read_paths[1]
        except IndexError:
            return None


def calculate_skewer_trimming_parameters(
    sample: WFSample,
    min_read_length: int,
) -> SkewerConfiguration:
    """Calculates trimming parameters based on the library type, and minimum allowed trim length.

    :param sample: The sample to calculate trimming parameters for.
    :param min_read_length: The minimum length of a read before it is discarded.
    :return: the trimming parameters
    """
    config = SkewerConfiguration(
        min_length=min_read_length,
        mode=SkewerMode.PAIRED_END if sample.paired else SkewerMode.SINGLE_END,
    )

    if sample.library_type == LibraryType.amplicon:
        config.end_quality = 0
        config.mean_quality = 0
        config.min_length = min_read_length

        return config

    if sample.library_type == LibraryType.srna:
        config.max_length = 22
        config.min_length = 20

        return config

    raise ValueError(f"Unknown library type: {sample.library_type}")


class SkewerRunner(Protocol):
    """A protocol describing callables that can be used to run Skewer."""

    async def __call__(
        self,
        config: SkewerConfiguration,
        paths: ReadPaths,
        output_path: Path,
    ) -> SkewerResult: ...


@fixture
def skewer(proc: int, run_subprocess: RunSubprocess) -> SkewerRunner:
    """Provides an asynchronous function that can run skewer.

    The provided function takes a :class:`.SkewerConfiguration` and a tuple of paths to
    the left and right reads to trim. If a single member tuple is provided, the dataset
    is assumed to be unpaired.

    The Skewer process will automatically be assigned the number of processes configured
    for the workflow run.

    Example:
    -------
    .. code-block:: python

        @step
        async def step_one(skewer: SkewerRunner, work_path: Path):
           config = SkewerConfiguration(
               mean_quality=30
           )

           skewer_result = await skewer(config, (
               work_path / "test_1.fq.gz",
               work_path / "test_2.fq.gz",
           ))


    """
    if shutil.which("skewer") is None:
        raise RuntimeError("skewer is not installed.")

    async def func(
        config: SkewerConfiguration,
        read_paths: ReadPaths,
        output_path: Path,
    ):
        temp_path = Path(await asyncio.to_thread(mkdtemp, suffix="_virtool_skewer"))

        await asyncio.to_thread(output_path.mkdir, exist_ok=True, parents=True)

        command = [
            str(a)
            for a in [
                "skewer",
                "-r",
                config.max_error_rate,
                "-d",
                config.max_indel_rate,
                "-m",
                config.mode.value,
                "-l",
                config.min_length,
                "-q",
                config.end_quality,
                "-Q",
                config.mean_quality,
                "-t",
                proc,
                # Skewer spams the console with progress updates. Set quiet to avoid.
                "--quiet",
                # Compress the trimmed output.
                "-z",
                "-o",
                f"{temp_path}/reads",
                *read_paths,
            ]
        ]

        process = await run_subprocess(
            command,
            cwd=read_paths[0].parent,
            env={**os.environ, "LD_LIBRARY_PATH": "/usr/lib/x86_64-linux-gnu"},
        )

        read_paths = await asyncio.to_thread(
            _rename_trimming_results,
            temp_path,
            output_path,
        )

        return SkewerResult(command, output_path, process, read_paths)

    return func


def _rename_trimming_results(temp_path: Path, output_path: Path) -> ReadPaths:
    """Rename Skewer output to a simple name used in Virtool.

    :param path: The path containing the results from Skewer
    """
    shutil.move(
        temp_path / "reads-trimmed.log",
        output_path / "trim.log",
    )

    try:
        return (
            shutil.move(
                temp_path / "reads-trimmed.fastq.gz",
                output_path / "reads_1.fq.gz",
            ),
        )
    except FileNotFoundError:
        return (
            shutil.move(
                temp_path / "reads-trimmed-pair1.fastq.gz",
                output_path / "reads_1.fq.gz",
            ),
            shutil.move(
                temp_path / "reads-trimmed-pair2.fastq.gz",
                output_path / "reads_2.fq.gz",
            ),
        )

File: apps/nuvs/src/skewer.ts
/**
 * Running skewer, ported from `utils.py`'s `skewer` fixture.
 *
 * Skewer is the one tool in this workflow with no runner anywhere else in the
 * repo, so the command and the output renaming both live here.
 *
 * `calculate_skewer_trimming_parameters` is **not** ported. It raises for every
 * library type but `srna` and nothing in the workflow calls it — porting it would
 * transliterate a function whose only reachable path is a `ValueError`.
 */

import { access, mkdir, mkdtemp, rename } from "node:fs/promises";
import { dirname, join } from "node:path";
import { matchToolVersion, type RunSubprocess } from "@virtool/workflow";

export const SKEWER_TOOL = "skewer";

/** How skewer is told to treat the reads it is given. */
export type SkewerMode = "pe" | "any";

/**
 * Skewer's defaults, as Python's `SkewerConfiguration` declares them.
 *
 * Every one is a field in the `trimmed_reads` cache key, so a change here forks
 * the namespace from Python nuvs'. `max_error_rate` and `max_indel_rate` are
 * Python **floats** and are marked as such where the params are built.
 */
export const SKEWER_END_QUALITY = 20;
export const SKEWER_MAX_ERROR_RATE = 0.1;
export const SKEWER_MAX_INDEL_RATE = 0.03;
export const SKEWER_MEAN_QUALITY = 25;

/**
 * The minimum length skewer keeps, from the sample's longest observed read.
 *
 * Python's `calculate_trimming_min_length`. The library type is a parameter
 * there and unread, so it is not one here.
 */
export function calculateTrimmingMinLength(maxLength: number): number {
	if (maxLength < 80) {
		return 35;
	}

	if (maxLength < 160) {
		return 100;
	}

	return 160;
}

/**
 * How `bowtie2-build` and most other tools announce themselves: a bare `version`
 * followed by the number.
 */
const LABELLED_VERSION = /\bversion\s+(\S+)/;

/**
 * The fallback: the first dotted number anywhere in the output.
 *
 * Skewer prints `skewer version: 0.2.2`, and the colon means
 * {@link LABELLED_VERSION} does not match — `\s+` wants whitespace immediately
 * after the word. Python tries the same two patterns in the same order for the
 * same reason, so dropping this reads the version as unparseable.
 */
const BARE_VERSION = /\bv?([0-9]+(?:\.[0-9A-Za-z_-]+)+)/;

/**
 * Read skewer's version.
 *
 * Both streams are collected, and a non-zero exit is **not** tolerated — unlike
 * `cd-hit-est`, whose help text is its only version output. Python does not
 * catch here either.
 */
export async function getSkewerVersion(
	runSubprocess: RunSubprocess,
): Promise<string> {
	const lines: string[] = [];

	const collect = (line: string) => {
		lines.push(line);
	};

	await runSubprocess({
		command: [SKEWER_TOOL, "--version"],
		stdout: collect,
		stderr: collect,
	});

	return matchToolVersion(
		LABELLED_VERSION.test(lines.join("\n")) ? LABELLED_VERSION : BARE_VERSION,
		lines,
		"Could not parse skewer version",
	);
}

/** What one skewer run needs. */
export type RunSkewerOptions = {
	minLength: number;
	mode: SkewerMode;
	/** Where `reads_1.fq.gz`, its pair and `trim.log` end up. */
	outputPath: string;
	proc: number;
	/** The sample's reads, in pair order. */
	readPaths: readonly string[];
	runSubprocess: RunSubprocess;
	/**
	 * Where skewer writes before its output is renamed.
	 *
	 * On the work path rather than the OS temp directory, which is what Python
	 * uses: the rename below is only atomic — and only free — within one
	 * filesystem, and the work path is the volume a pod is sized for.
	 */
	stagingParent: string;
};

/**
 * Trim `readPaths` into `outputPath`.
 *
 * The flag order is Python's verbatim. It has no effect on skewer's behaviour,
 * but the command is the thing a failed run is debugged from and a reordered one
 * is needlessly hard to diff against Python's.
 */
export async function runSkewer({
	minLength,
	mode,
	outputPath,
	proc,
	readPaths,
	runSubprocess,
	stagingParent,
}: RunSkewerOptions): Promise<void> {
	const [firstRead] = readPaths;

	if (firstRead === undefined) {
		throw new Error("Cannot run skewer with no reads");
	}

	await mkdir(stagingParent, { recursive: true });
	await mkdir(outputPath, { recursive: true });

	const staging = await mkdtemp(join(stagingParent, "skewer-"));

	await runSubprocess({
		command: [
			SKEWER_TOOL,
			"-r",
			String(SKEWER_MAX_ERROR_RATE),
			"-d",
			String(SKEWER_MAX_INDEL_RATE),
			"-m",
			mode,
			"-l",
			String(minLength),
			"-q",
			String(SKEWER_END_QUALITY),
			"-Q",
			String(SKEWER_MEAN_QUALITY),
			"-t",
			String(proc),
			// Skewer spams stdout with progress updates.
			"--quiet",
			// Compress the trimmed output.
			"-z",
			"-o",
			join(staging, "reads"),
			...readPaths,
		],
		// Python runs skewer from the reads' own directory. Nothing in the command
		// is relative, so this only decides where skewer would drop a stray file.
		cwd: dirname(firstRead),
		// Skewer in the tools image is linked against libraries the runtime base
		// keeps here rather than on the default search path.
		env: { LD_LIBRARY_PATH: "/usr/lib/x86_64-linux-gnu" },
	});

	await renameTrimmingResults(staging, outputPath);
}

async function exists(path: string): Promise<boolean> {
	try {
		await access(path);

		return true;
	} catch {
		return false;
	}
}

/**
 * Move skewer's output to the names the rest of the workflow uses.
 *
 * **Whether the run was paired is read off the output, not off the sample.**
 * Skewer names its files by what it did — `reads-trimmed.fastq.gz` for one input,
 * `reads-trimmed-pair{1,2}.fastq.gz` for two — and that is what Python branches
 * on, by catching the `FileNotFoundError` from the single-end name. Tested here
 * rather than caught, so a genuinely absent paired file still fails naming
 * itself.
 */
async function renameTrimmingResults(
	staging: string,
	outputPath: string,
): Promise<void> {
	await rename(
		join(staging, "reads-trimmed.log"),
		join(outputPath, "trim.log"),
	);

	const singleEnd = join(staging, "reads-trimmed.fastq.gz");

	if (await exists(singleEnd)) {
		await rename(singleEnd, join(outputPath, "reads_1.fq.gz"));

		return;
	}

	await rename(
		join(staging, "reads-trimmed-pair1.fastq.gz"),
		join(outputPath, "reads_1.fq.gz"),
	);

	await rename(
		join(staging, "reads-trimmed-pair2.fastq.gz"),
		join(outputPath, "reads_2.fq.gz"),
	);
}
